pkg/globalnet/controllers/ipam/gatewaymonitor.go

```diff
@@ -124,197 +127,104 @@ func (gm *GatewayMonitor) Stop() {
 }
 
-func (gm *GatewayMonitor) runEndpointWorker() {
-	for gm.processNextEndpoint() {
-	}
-}
-
-func (gm *GatewayMonitor) processNextEndpoint() bool {
-	obj, shutdown := gm.endpointWorkqueue.Get()
-	if shutdown {
-		return false
-	}
-
-	err := func() error {
-		defer gm.endpointWorkqueue.Done(obj)
-		key := obj.(string)
-		ns, name, err := cache.SplitMetaNamespaceKey(key)
-		if err != nil {
-			gm.endpointWorkqueue.Forget(obj)
-			return fmt.Errorf("error while splitting meta namespace key %s: %v", key, err)
-		}
-
-		endpoint, err := gm.submarinerClientSet.SubmarinerV1().Endpoints(ns).Get(context.TODO(), name, metav1.GetOptions{})
-		if err != nil {
-			gm.endpointWorkqueue.Forget(obj)
-			return fmt.Errorf("error retrieving submariner endpoint object %s: %v", name, err)
-		}
-
-		klog.V(log.DEBUG).Infof("In processNextEndpoint, endpoint info: %+v", endpoint)
-
-		if endpoint.Spec.ClusterID != gm.clusterID {
-			klog.V(log.DEBUG).Infof("Endpoint %s belongs to a remote cluster", endpoint.Spec.Hostname)
-
-			overlap, err := cidr.IsOverlapping(endpoint.Spec.Subnets, gm.ipamSpec.GlobalCIDR[0])
-			if err != nil {
-				// Ideally this case will never hit, as the subnets are valid CIDRs
-				klog.Warningf("unable to validate overlapping Service CIDR: %s", err)
-			}
-
-			if overlap {
-				// When GlobalNet is used, globalCIDRs allocated to the clusters should not overlap.
-				// If they overlap, skip the endpoint as its an invalid configuration which is not supported.
-				klog.Errorf("GlobalCIDR %q of local cluster %q overlaps with remote cluster %s",
-					gm.ipamSpec.GlobalCIDR[0], gm.ipamSpec.ClusterID, endpoint.Spec.ClusterID)
-				gm.endpointWorkqueue.Forget(obj)
+func (gm *GatewayMonitor) handleCreatedOrUpdatedEndpoint(obj runtime.Object, numRequeues int) bool {
+	endpoint := obj.(*v1.Endpoint)
 
-				return nil
-			}
+	klog.V(log.DEBUG).Infof("In processNextEndpoint, endpoint info: %+v", endpoint)
 
-			for _, remoteSubnet := range endpoint.Spec.Subnets {
-				MarkRemoteClusterTraffic(gm.ipt, remoteSubnet, AddRules)
-			}
+	if endpoint.Spec.ClusterID != gm.clusterID {
+		klog.V(log.DEBUG).Infof("Endpoint %s belongs to a remote cluster", endpoint.Spec.Hostname)
 
-			gm.endpointWorkqueue.Forget(obj)
-
-			return nil
-		}
-
-		endpoints, err := gm.submarinerClientSet.SubmarinerV1().Endpoints(ns).List(context.TODO(), metav1.ListOptions{})
+		overlap, err := cidr.IsOverlapping(endpoint.Spec.Subnets, gm.ipamSpec.GlobalCIDR[0])
 		if err != nil {
-			gm.endpointWorkqueue.Forget(obj)
-			return fmt.Errorf("error retrieving submariner endpoint list %v", err)
+			// Ideally this case will never hit, as the subnets are valid CIDRs
+			klog.Warningf("unable to validate overlapping Service CIDR: %s", err)
 		}
 
-		for _, endpoint := range endpoints.Items {
-			if endpoint.Spec.ClusterID != gm.clusterID {
-				for _, remoteSubnet := range endpoint.Spec.Subnets {
-					MarkRemoteClusterTraffic(gm.ipt, remoteSubnet, AddRules)
-				}
-			}
-		}
```

Comment:
@tpantelis I was debugging the issue with Wireguard driver and found the problem. The following code is accidentally removed in the new implementation.

I was wondering why it was working with Libreswan Cabledriver and noticed that there is a difference of behavior between the two. In Libreswan driver, when the Gateway pod is restarted, it was not deleting the local Endpoint but with Wireguard driver, it was deleting the local endpoint. Anyways, IMHO I feel we should fix the code in Globalnet.

Reply:
Thanks for looking into this. I remember this code now - it didn't seem necessary since we do the same thing above on L173 when we get notified of the remote Endpoint. So I'm not clear on why we would also do this when for a local Endpoint, ie it seems redundant.

Reply:
Fixed with https://github.com/submariner-io/submariner/pull/1196/commits/63cc50ef87f7a04d3b8e3ab47c7fc7be7ed8cd66.

Reply:
Thanks for fixing it @tpantelis